### ISSUE TITLE
Add JUnit XML reporting

### ISSUE THREAD

**Issue by simonvanderveldt** [CodeNarc/CodeNarc#189]
Would it be possible/align with CodeNarcs goals to add a simple JUnit XML report writer?

That way it would be possible to integrate the results with tools that can process/aggregate JUnit XML files, which entails a lot more tools than the small list of tools that can process the XML report as it's currently generated by CodeNarc.

Our use case is we'd like some simple reporting or static analysis/linting using CodeNarc in our CI, which can process JUnit XMLs.

**Comment by chrismair** [comment]
Interesting request. There does seem to be something of an _impedance mismatch_ between a CodeNarc report and a JUnit test report.

I found this as one source of info about the JUnit format, and it includes several other links:
http://stackoverflow.com/questions/4922867/junit-xml-format-specification-that-hudson-supports

Would a _testcase_ map to a file? And a rule violation is a _failure_ or _error?_ But then it can only have a single file/error, rather than multiple violations?

**Comment by simonvanderveldt** [comment]
> Interesting request. There does seem to be something of an impedance mismatch between a CodeNarc report and a JUnit test report.

Good point. I just compared the formats andI think it's possible to make it match by simply reporting the issues encountered per package. Which means there'd be 0 items in the test report if everything is OK. This model matches how JUnit test results are reported.

> I found this as one source of info about the JUnit format, and it includes several other links:
> http://stackoverflow.com/questions/4922867/junit-xml-format-specification-that-hudson-supports

AFAIK the JUnit spec XML spec actually isn't a JUnit thing, but an ant thing. The source for the XML generator lives [here](https://github.com/apache/ant/blob/master/src/main/org/apache/tools/ant/taskdefs/optional/junit/XMLJUnitResultFormatter.java)
I believe the matching schema definition is [this one from surefire](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/site/resources/xsd/surefire-test-report.xsd).
Also, [this page](https://www.ibm.com/support/knowledgecenter/SSQ2R2_9.5.0/com.ibm.rsar.analysis.codereview.cobol.doc/topics/cac_useresults_junit.html) gives a pretty nice overview.

Compared to the HTML report this is what I think the relations are:
- `testsuite` = package
- `testcase` = priority: rule name
- `testcase` property/field `classname` = classname ;)
  - `failure` = Source Line / Message

Sample output would look something like this:
```
<?xml version="1.0"?>
<testsuite name="CodeNarc" package="foo" timestamp="2017-02-03T10:00:00" tests="3" errors="0" failures="3" time="2.000">
  <testcase name="2: LineLength" time="0.01" classname="bar">
    <failure message="bar.java:100: LineLength">The line exceeds 120 characters. The line is 1337 characters.</failure>
  </testcase>
  <testcase name="3: UnusedImport" time="0.01" classname="bar">
    <failure message="bar.java:3: UnusedImport">The [org.someorg.someclass] import is never referenced</failure>
  </testcase>
  <testcase name="3: UnusedImport" time="0.01" classname="baz">
    <failure message="baz.java:5: UnusedImport">The [org.someorg.someclass] import is never referenced</failure>
  </testcase>
</testsuite>
```

**Comment by chrismair** [comment]
One wrinkle is that a rule violation can occur multiple times per file. Another is that some rule violations are associated with the file rather than a specific class (that's why the XML report is by file and not by class).

**Comment by simonvanderveldt** [comment]
@chrismair thanks for thinking with me/the feedback! :)

> One wrinkle is that a rule violation can occur multiple times per file

I don't think this is really a problem, since the XML and parsers don't care what kind of failures/test results are in the list and also don't care if there are duplicates.
Also when adding the line number to the failure message they would all be clearly distinguishable.

> Another is that some rule violations are associated with the file rather than a specific class (that's why the XML report is by file and not by class).

If that's the unit that's used for grouping in CodeNarc then I think the simplest thing to do is just use the filename for grouping ;) I.e. fill in the filename in the `classname` field.
I've updated the sample above to reflect this

**Comment by edupo** [comment]
What about one JUnit testcase could look like:
```assert: number of violations this rule == 0```
So the JUnit always contain the same set of tests but those pass or fail depending on the violated rule.
The stderr of the test can be the list of files that violates the rule.

I don't know if this logic belongs to CodeNarc though.

**Comment by keegangroth** [comment]
I wouldn't mind having this. In the ruby world, rubocop handles with https://docs.rubocop.org/rubocop/formatters.html#junit-style-formatter (it also has the handy option to generate multiple formats)

**Comment by janisz** [comment]
In golang world there is also tiple format supported from linter https://golangci-lint.run/usage/configuration/#output-configuration

**Comment by marmstrong-cis** [comment]
I agree that this would add a good bit of functionality for our project. Right now, we build in AWS, and they allow for reports to be uploaded. [Here is a link of supported types](https://docs.aws.amazon.com/codebuild/latest/userguide/test-reporting.html), but I think JUNIT might be the most beneficial for an MVP
